stock_inventory_revaluation/models/stock_inventory_revaluation.py

```diff
@@ -0,0 +1,479 @@
+        stock_valuation_account_id = False
+        if datas.get('stock_valuation'):
+            stock_valuation_account_id = datas.get('stock_valuation').id
+        if not stock_valuation_account_id:
+            raise UserError(_("Please add Stock Valuation Account in "
+                            "Product Category"))
+        if not self.decrease_account_id or not self.increase_account_id:
+            raise UserError(_("Please add an Increase Account and "
+                              "a Decrease Account."))
+        prec = self.env['decimal.precision'].precision_get('Account')
+        for prod_variant in self.product_id:
+            amount_diff = 0.0
+            if self.product_id.cost_method == 'real':
+                for reval_quant in self.reval_quant_ids:
+                    if reval_quant.product_id == prod_variant:
+                        amount_diff += reval_quant.get_total_value()
+                if amount_diff == 0.0:
+                    return True
+            else:
+                if self.revaluation_type == 'price_change':
+                    diff = self.old_cost - self.new_cost
+                    amount_diff = prod_variant.qty_available * diff
+                else:
+                    proportion = prod_variant.qty_available / \
+                        self.product_template_id.qty_available
+                    amount_diff = round(self.new_value * proportion, prec)
+
+            qty = prod_variant.qty_available
+            if qty:
+                if amount_diff > 0:
+                    debit_account_id = self.decrease_account_id.id
+                    credit_account_id = \
+                        datas.get('stock_valuation').id
+
+                else:
+                    debit_account_id = \
+                        datas.get('stock_valuation').id
+                    credit_account_id = self.increase_account_id.id
+                move = self.env['account.move'].create(move_data)
+                move_line_data = self._prepare_debit_move_line_data(
+                    move, abs(amount_diff), debit_account_id, prod_variant.id)
+                move_line_obj.with_context({'check_move_validity':
+                                            False}).create(move_line_data)
+                move_line_data = self._prepare_credit_move_line_data(
+                    move, abs(amount_diff), credit_account_id, prod_variant.id)
+                move_line_obj.create(move_line_data)
+                move.post()
+
+    @api.multi
+    def post(self):
+        for revaluation in self:
+            if revaluation.product_id.\
+                    cost_method == 'real':
+                for reval_quant in revaluation.reval_quant_ids:
+                    reval_quant.old_cost = reval_quant.quant_id.cost
+                    reval_quant.write_new_cost()
+            else:
+                if revaluation.product_id.\
+                        cost_method in ['standard', 'average']:
```

Comment:
I'm a bit confused.  Should standard and average costing methods be allowed here at all if the assumption is to use  "Set standard price" link in product form for these?

If these are allowed, I think `button_cancel` method should also take care of reverting the standard price of the product template.

Reply:
@yostashiro When you use the "set standard price" link, it creates a revaluation. So why not allow it to be used here?

As for reverting the standard price, it is not straightforward, because the average price may have already changed very much, due to new purchases.

Reply:
@jbeficent 

> When you use the "set standard price" link, it creates a revaluation. So why not allow it to be used here?

Because README states that:
> For products set with average or standard price and real-time valuation, go to the Product form and use the "Set standard price" link to change the standard price.

And cancellation logic (`button_cancel`) at the moment doesn't support the revert for average and standard costing methods properly.  Therefore, the current design appears to me a bit confusing/inconsistent. 

Some options I can think of:

Option 1. Make average and standard costing methods out of scope of this module.

Option 2. Keep the current design, but update README to mention the limitation.  Preferably hide "Cancel" button for products with average/standard costing method.

Option 3. Update `button_cancel` to handle update of `standard_price` of the product template.  I think this option can be tricky as you say especially for average method, but I guess it may not be so difficult to allow cancellation only when there is no subsequent changes to `standard_price` after the revaluation operation.  e.g. add check against product.price.history records.  It's still not perfect, but sounds reasonable/acceptable to me.

Reply:
Let's go for option 2, and hide the "Cancel" button in these cases cc @mreficent